[pytorch] Expand PixelShuffle to support any number of batch dims

jbschlosser opened the pull request:
Summary: Expands the implementation of PixelShuffle to support any number of batch dimensions. This addresses https://github.com/pytorch/pytorch/issues/42150

Test Plan: `buck test caffe2/test:nn -- test_pixel_shuffle`

Differential Revision: D25399058

mruberry commented on aten/src/ATen/native/PixelShuffle.cpp:
```diff
@@ -4,31 +4,45 @@
-  int64_t h = self.size(2);
-  int64_t w = self.size(3);
+  TORCH_CHECK(self.dim() >= 3,
+             "pixel_shuffle expects at least 3D input, but got input with sizes ", self.sizes());
```
This error message is not a regression, but I think it could be slightly improved.

First, the check is for dimensionality, but when this check fails the sizes are shown to the user. It seems like it'd be more consistent to say the input had X dims to be consistent with the check.

Second, the use of "3D" here may be misleading, as pixel shuffle shuffles batches of (conceptually) 3D tensors. I think this can also be addressed by consistently referencing dims, so the error message would be something like:

```
"... expects input to have at least 3 dimensions, but got input with X dimensions"
```

mruberry commented on aten/src/ATen/native/PixelShuffle.cpp:
```diff
@@ -4,31 +4,45 @@
   int64_t upscale_factor_squared = upscale_factor * upscale_factor;
-  AT_ASSERTM(c % upscale_factor_squared == 0,
+  TORCH_CHECK(c % upscale_factor_squared == 0,
              "pixel_shuffle expects input channel to be divisible by square of "
```
"expects input channel to be divisible by square of" -> "expects its input's 'channel' dimension to be divisible by the square of"

mruberry commented on aten/src/ATen/native/PixelShuffle.cpp:
```diff
@@ -4,31 +4,45 @@
-  AT_ASSERTM(c % upscale_factor_squared == 0,
+  TORCH_CHECK(c % upscale_factor_squared == 0,
              "pixel_shuffle expects input channel to be divisible by square of "
              "upscale_factor, but got input with sizes ", self.sizes(),
```
the construction here is a little awkward:

"..., but got input with sizes X, upscale_factor=Y, and self.size(-3)=Z is not divisible by A"

Doubly so since the name "self" is not visible to the user in this context.

One fix would be to simplify the message to:

"...but input.size(-3)=X is not divisible by Y".

mruberry commented on test/test_nn.py:
```diff
@@ -6519,18 +6519,32 @@ def test_noncontig_conv_grad_cuda(self, dtype=torch.float):
         self.assertEqual(result, input.grad.data, atol=dtype2prec_DONTUSE[dtype], rtol=0)
 
     def test_pixel_shuffle(self):
```
This test looks good; is there also a test for the two error cases? If not, let's add one to verify the module throws both those exceptions as expected.

mruberry commented on test/test_nn.py:
```diff
@@ -6519,18 +6519,32 @@ def test_noncontig_conv_grad_cuda(self, dtype=torch.float):
+            input = torch.rand(*batch_sizes, channels, height, width, requires_grad=True)
+            ps = nn.PixelShuffle(upscale_factor)
+            output = ps(input)
+            self._verify_pixel_shuffle(input.data, output.data, upscale_factor)
```
What's up with the use of .data here?

mruberry commented on test/test_nn.py:
```diff
@@ -6519,18 +6519,32 @@ def test_noncontig_conv_grad_cuda(self, dtype=torch.float):
+            output = ps(input)
+            self._verify_pixel_shuffle(input.data, output.data, upscale_factor)
+            output.backward(output.data)
+            self.assertEqual(input.data, input.grad.data)
```
This is an interesting autograd test. I don't think this PR needs to change it, though.

mruberry commented on aten/src/ATen/native/PixelShuffle.cpp:
```diff
@@ -4,31 +4,45 @@
-                       .reshape({b, oc, oh, ow});
+  // Subtract the 3 non-batch dims (C, H, W).
+  const auto last_batch_dim = self.sizes().end() - 3;
+  std::vector<int64_t> upscaled_shape(self.sizes().begin(), last_batch_dim);
```
This is a really cool algorithm and it's not immediately obvious how it works. Add a comment or two here explaining the approach and what each of its 3 steps does.

jbschlosser replied:
Good point, it's fairly cryptic as-is. I took a pass at commenting the flow of the 3 steps- let me know what you think of the description and I can tune it further if it's still unclear.

mruberry commented on torch/nn/modules/pixelshuffle.py:
```diff
@@ -15,12 +15,15 @@ class PixelShuffle(Module):
     `Real-Time Single Image and Video Super-Resolution Using an Efficient Sub-Pixel Convolutional Neural Network`_
     by Shi et. al (2016) for more details.
 
+    Note that this function can take inputs with any number of batch dimensions:
```
TODO: not for this PR but we should update the PixelShuffle and pixel_shuffle documentation. They start off by referring to a value named "r" for example, without ever defining what "r" is!

mruberry commented on test/test_nn.py:
```diff
@@ -6480,14 +6480,14 @@ def test_RNN_change_dropout(self):
                 prev_output = output1.data
 
     def _verify_pixel_shuffle(self, input, output, upscale_factor):
```
Add a comment explaining what this function does. If it's only used inside `_test_pixel_shuffle_helper` then it can be moved inside that function.

mruberry reviewed (commented):
Hey @jbschlosser! Overall this looks great. I made a few comments about test coverage, error messages, and comments.

jbschlosser committed:
[pytorch] Expand PixelShuffle to support any number of batch dims (#49187)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49187

Expands the implementation of PixelShuffle to support any number of batch dimensions

Test Plan: `buck test caffe2/test:nn -- test_pixel_shuffle`

Differential Revision: D25399058

fbshipit-source-id: c1a36d17660ef3dce0cdfb4ec419f54124d7b809

mruberry reviewed (approved):
Nice work, @jbschlosser!

Some interesting possible follow-up tasks so we don't forget:

- pixel shuffle's documentation could be revised
- its autograd testing could be improved
- should this function ultimately be in torchvision and not torch?
